### PR TITLE
Perf (opt‑in): Mixed‑precision p_cmp logits via CUDA autocast (NSA_P_CMP_MIXED=1)

### PR DESCRIPTION
**Pull request by seconds-0**
Title: Perf (opt‑in): Mixed‑precision p_cmp logits via CUDA autocast (NSA_P_CMP_MIXED=1)

Summary
- Adds an OPT‑IN mixed‑precision path for selection scoring logits `p_cmp` on CUDA:
  - Under `NSA_P_CMP_MIXED=1`, `compute_pcmp_all(...)` runs einsum+softmax inside `torch.autocast(device_type='cuda', dtype=torch.bfloat16)`.
  - Output upcasts back to the original dtype to preserve downstream numerics.
- Default OFF; no behavioral change without the flag.

Branch & Commit
- Branch: perf/pcmp-autocast
- Head: 569e0f58

Motivation
- `p_cmp` dominates selection scoring bandwidth at large (S,G,h). Autocast reduces memory traffic and kernel time on A100/H100 while keeping accuracy/tie‑break semantics (we upcast before top‑k).

Implementation
- nsa/core/selection_scorer.py
  - `compute_pcmp_all(...)` checks `NSA_P_CMP_MIXED`. If enabled and `device='cuda'`, wraps logits+softmax in autocast(bf16), then returns `p.to(orig_dtype)`.
  - Baseline precise path unchanged when flag is off.

GPU Test Plan

Setup
```bash
uv venv -p 3.11 .venv && source .venv/bin/activate
uv pip sync -r requirements-gpu-cu121-torch24.txt
# Optional FA‑2; not required for this PR
pip install flash-attn --no-build-isolation || true

# Checkout branch
git fetch origin --prune
git checkout perf/pcmp-autocast && git reset --hard 569e0f58
```

Core env
```bash
export NSA_PREFILL_BATCHED=1
export NSA_USE_SEL_PACK=1
export NSA_SEL_RANGES_V2=1
export NSA_USE_FA2=1 NSA_USE_FA2_CMP=1 NSA_USE_FA2_WIN=0
```

A/B runs
1) Baseline (mixed off)
```bash
unset NSA_P_CMP_MIXED
# Correctness
PYTHONPATH=. uv run -q pytest -q nsa/tests/test_selection_v2_equiv.py
PYTHONPATH=. uv run -q pytest -q nsa/tests/test_equiv_small.py nsa/tests/test_group_consistency.py nsa/tests/test_masks.py nsa/tests/test_block_math.py
# Benchmarks
PYTHONPATH=. uv run python bench/bench_decode.py --S_list 512,1024,2048
CONFIG=configs/base.yaml PYTHONPATH=. uv run -q python -u scripts/train_showcase.py --dataset fineweb_edu --ddp 0 --steps 200
```
2) Mixed precision ON
```bash
export NSA_P_CMP_MIXED=1
# Repeat the same battery
PYTHONPATH=. uv run -q pytest -q nsa/tests/test_selection_v2_equiv.py
PYTHONPATH=. uv run -q pytest -q nsa/tests/test_equiv_small.py nsa/tests/test_group_consistency.py nsa/tests/test_masks.py nsa/tests/test_block_math.py
PYTHONPATH=. uv run python bench/bench_decode.py --S_list 512,1024,2048
CONFIG=configs/base.yaml PYTHONPATH=. uv run -q python -u scripts/train_showcase.py --dataset fineweb_edu --ddp 0 --steps 200
```

What to look for
- All tests green in both modes; no NaNs; selection v2 parity intact.
- Decode bench and 200‑step training tokens/sec improved or unchanged with `NSA_P_CMP_MIXED=1`.
- No behavior changes when the flag is OFF (default).

Acceptance
- No correctness/stability regressions.
- Non‑regressing (ideally improved) throughput at larger S with the flag enabled.

Notes
- This PR is independent of selection varlen PR. It complements varlen selection by cutting bandwidth earlier in the selection scoring pipeline.

### CODE REVIEW

**Comment by seconds-0** [comment]
Closing this PR as it reintroduces critical bugs that were already fixed in PR #26:

1. **Critical indentation bug**: SDPA computation incorrectly nested in else block (lines 367-386)
2. **Removes safety bounds**: Deletes _env_int_bounded() function that prevents excessive memory allocation
3. **Duplicate commits**: All commits are already in master (workspace pre-sizing, remove repeat_interleave, mixed-precision)
4. **Incorrect documentation**: Includes unrelated validation report

The mixed-precision feature this PR intended to add is already merged via PR #28/29 (commit f1c43044).

This appears to be based on an outdated branch that predates the fixes from PR #26.
